### PR TITLE
Phase 2: Noise-encrypted P2P transport

### PR DESCRIPTION
**Pull request by gdassori**
## Summary

- Noise XX encrypted TCP connections (X25519, ChaChaPoly, BLAKE2s)
- Protobuf wire protocol: `Envelope` wrapper with ED25519 signatures, UUID-based deduplication
- PeerHello handshake with 3-check identity verification (node_id derivation, X25519 key binding, envelope signature)
- Bidirectional chat relay between partyline hubs via transport manager fanout
- Outbound-only nodes (no listen port) are first-class citizens
- Exponential backoff reconnection for bootstrap peers
- MkDocs Material protocol documentation + GitHub Pages deployment workflow

## New packages

| Package | Purpose |
|---------|---------|
| `internal/crypto` | ED25519 → X25519 key derivation |
| `internal/transport` | Noise handshake, wire framing (`0x4D49` magic, 1MB max), peer lifecycle, connection manager |
| `proto/` | Protobuf definitions (Envelope, PeerHello, PeerHelloAck, ChatMessage) |
| `pkg/proto/` | Generated Go code |
| `docs/` | Protocol specification (identity, transport, messages, connection lifecycle, architecture) |

## Modified

- `internal/partyline` — `RemoteMsg`, `SetRemoteSend()`, `DeliverRemote()` for hub↔transport bridge
- `cmd/micelio/main.go` — transport manager wiring (optional, only when `network.listen` is set)
- `Makefile` — `proto` target
- `go.mod` — new deps: `flynn/noise`, `filippo.io/edwards25519`, `google.golang.org/protobuf`, `google/uuid`

## Test plan

- [x] `TestEdToX25519RoundTrip` — key conversion correctness
- [x] `TestFrameRoundTrip` + 4 edge cases — wire framing
- [x] `TestTwoNodeChat` — bidirectional A↔B chat
- [x] `TestTenNodeStar` — 10-node star: center→spokes, spoke→center, no cross-spoke gossip
- [x] `TestOutboundBridge` — outbound-only bridge joining two 3-node networks: bridge↔both centers, no cross-network leak, spoke isolation
- [x] `TestSSHPartyline` — existing SSH integration test still passes
- [x] All 14 tests pass locally (`go test ./...`)

### CODE REVIEW

File: internal/transport/noise.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	var lenBuf [2]byte
+	binary.BigEndian.PutUint16(lenBuf[:], uint16(len(ciphertext)))
+
+	if _, err := nc.conn.Write(lenBuf[:]); err != nil {
+		return 0, err
+	}
+	if _, err := nc.conn.Write(ciphertext); err != nil {
+		return 0, err
+	}
+	return len(p), nil
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/noise.go]: Fixed in b2bee32. Upgraded Noise transport framing from uint16 to uint32 length prefix.

File: internal/partyline/partyline.go [hunk 2]
```diff
@@ -16,6 +22,8 @@ type Hub struct {
 	who       chan whoReq
 	setNick   chan nickReq
 	stop      chan struct{}
+
+	remoteSend chan<- RemoteMsg // set by transport; nil if no P2P
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/partyline/partyline.go]: Fixed in b2bee32. Reordered main.go: Manager is created (calling SetRemoteSend) before `go hub.Run()`.

File: internal/transport/manager_test.go [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+			nodes[i].cancel()
+			nodes[i].hub.Stop()
+		}
+	}()
+
+	// Start node 0 (center) with a listener
+	cfg0 := &config.Config{
+		Node:    config.NodeConfig{Name: "node-0", DataDir: nodes[0].id.NodeID},
+		Network: config.NetworkConfig{Listen: "127.0.0.1:0"},
+	}
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/manager_test.go]: WONTFIX. DataDir is not used by the transport Manager. This is cosmetic.

File: internal/transport/manager_test.go [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+	hubB := partyline.NewHub("node-b")
+	go hubB.Run()
+	defer hubB.Stop()
+
+	// Node B listens
+	cfgB := &config.Config{
+		Node:    config.NodeConfig{Name: "node-b", DataDir: dirB},
+		Network: config.NetworkConfig{Listen: "127.0.0.1:0"},
+	}
+	mgrB, err := transport.NewManager(cfgB, idB, hubB)
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/manager_test.go]: Fixed in 005d61e. All tests now create Manager (which calls `SetRemoteSend`) before `go hub.Run()`, matching the pattern in `main.go`. Also protected `m.listener` with the mutex to eliminate the race detected by `-race` between `Start()` and `Addr()`.

File: cmd/micelio/main.go [hunk 5]
```diff
@@ -64,18 +65,44 @@ func main() {
...
+	defer cancel()
+
+	// Create transport manager before hub.Run() to avoid data race on remoteSend.
+	// Manager is created when listen OR bootstrap is configured (supports outbound-only nodes).
+	var mgr *transport.Manager
+	if cfg.Network.Listen != "" || len(cfg.Network.Bootstrap) > 0 {
+		var err2 error
+		mgr, err2 = transport.NewManager(cfg, id, hub)
+		if err2 != nil {
+			log.Fatalf("transport: %v", err2)
```
[the remark this answers is not in the document]
Comment (gdassori) [reply cmd/micelio/main.go]: WONTFIX — `err2` is a standard Go pattern when `err` is already in scope from a prior `:=` in the same function. The surrounding context (`transport.NewManager` + `log.Fatalf("transport: ...")`) makes the purpose clear. Renaming to `transportErr` adds verbosity without meaningful clarity.

File: internal/transport/frame.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	buf := make([]byte, HeaderSize+len(payload))
+	binary.BigEndian.PutUint16(buf[0:2], Magic)
+	binary.BigEndian.PutUint16(buf[2:4], Version)
+	binary.BigEndian.PutUint32(buf[4:8], uint32(len(payload)))
+	copy(buf[8:], payload)
+
+	_, err := w.Write(buf)
+	return err
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/frame.go]: WONTFIX — The only caller passes a `noiseConn` which wraps `net.Conn`. Go's `net.Conn.Write` handles partial writes internally (loops until all bytes sent or error). Adding a write loop here would be dead code.

File: internal/transport/manager.go [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+		select {
+		case <-ticker.C:
+			// reconnection is handled by dialWithBackoff goroutines
+		case <-ctx.Done():
+			return
+		case <-m.done:
+			return
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/manager.go]: Fixed in eadaa4b. `reconnectLoop` removed.

File: internal/transport/peer.go [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+func (p *Peer) handleChat(env *pb.Envelope) {
+	var chat pb.ChatMessage
+	if err := proto.Unmarshal(env.Payload, &chat); err != nil {
+		log.Printf("peer %s: unmarshal chat: %v", p.NodeID, err)
+		return
+	}
+	p.hub.DeliverRemote(chat.Nick, chat.Text)
+}
+
+func (p *Peer) sendHello(hello *pb.PeerHello, msgType uint32) error {
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/peer.go]: WONTFIX — Intentional design. `PeerHelloAck` is structurally identical to `PeerHello` by design; the `msg_type` field in the Envelope differentiates them. This avoids proto duplication while keeping the wire protocol explicit about directionality.

File: internal/transport/peer.go [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+
+	envBytes, err := proto.Marshal(env)
+	if err != nil {
+		return fmt.Errorf("marshal envelope: %w", err)
+	}
+
+	return WriteFrame(p.conn, envBytes)
+}
+
+func (p *Peer) recvHello(expectedType uint32) (*pb.PeerHello, error) {
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/peer.go]: WONTFIX — See reply on the `sendHello` comment above.

File: .github/workflows/docs.yml [hunk 10]
```diff
@@ -0,0 +1,48 @@
+name: Docs
+
+on:
+  push:
+    branches: [master]
```
[the remark this answers is not in the document]
Comment (gdassori) [reply .github/workflows/docs.yml]: WONTFIX — The repo uses `master` as the default branch (matching `ci.yml`). This is correct.

File: internal/transport/manager_test.go [hunk 11]
```diff
@@ -0,0 +1,587 @@
...
+			nodes[i].cancel()
+			nodes[i].hub.Stop()
+		}
+	}()
+
+	// Create manager for node 0 (center) before hub.Run() to avoid data race on remoteSend.
+	cfg0 := &config.Config{
+		Node:    config.NodeConfig{Name: "node-0", DataDir: nodes[0].id.NodeID},
+		Network: config.NetworkConfig{Listen: "127.0.0.1:0"},
+	}
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/manager_test.go]: WONTFIX — Cosmetic. `DataDir` is not used by the transport layer; it's only present in the config struct for completeness. No functional impact.

File: internal/crypto/keys_test.go [hunk 12]
```diff
@@ -0,0 +1,42 @@
+package crypto_test
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"testing"
+
+	"golang.org/x/crypto/curve25519"
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/crypto/keys_test.go]: WONTFIX — `x/crypto/curve25519` is already a transitive dependency (used by `flynn/noise`), so no extra dep is added. Switching to `crypto/ecdh` in the test would save no imports in `go.sum` and the current code is clearer for the purpose of the test (explicit scalar-mult).

File: internal/transport/noise.go [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	// Read one encrypted message
+	var lenBuf [4]byte
+	if _, err := io.ReadFull(nc.conn, lenBuf[:]); err != nil {
+		return 0, err
+	}
+	msgLen := binary.BigEndian.Uint32(lenBuf[:])
+
+	// Cap at MaxPayload + HeaderSize + 16 (Poly1305 tag) to prevent OOM from malicious length.
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/noise.go]: Fixed in 191ee7c. Improved the comment wording.

File: internal/transport/noise.go [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+	var lenBuf [2]byte
+	binary.BigEndian.PutUint16(lenBuf[:], uint16(len(msg)))
+	if _, err := w.Write(lenBuf[:]); err != nil {
+		return fmt.Errorf("noise handshake write len: %w", err)
+	}
+	if _, err := w.Write(msg); err != nil {
+		return fmt.Errorf("noise handshake write msg: %w", err)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (gdassori) [reply internal/transport/noise.go]: Fixed the valid part in 667fc86: added explicit `len(msg) > 0xFFFF` guard before the uint16 cast. The partial-write concern remains WONTFIX per previous reply (net.Conn guarantees full writes).
